### PR TITLE
Document how to parameterize output directory for REST Docs with WebTestClient

### LINKED ISSUE #27747: Document how to use WebTestClientBuilderCustomizer to configure REST Docs' parameterized output
We already have an example showing how to do it with `MockMvc`. It'd be good to have an equivalent example for `WebTestClient`. See https://github.com/spring-projects/spring-restdocs/issues/741 for some background.

### PR DESCRIPTION
**Pull request by brneto**
closes #27747

### CODE REVIEW

**Comment by brneto** [comment]
@wilkinsona It failed with message `A build scan was not published as you have not authenticated with server 'ge.spring.io'.`, is there something more I have to do?

File: spring-boot-project/spring-boot-docs/src/docs/asciidoc/features/testing.adoc [hunk 1]
```diff
@@ -770,6 +779,15 @@ If you require more control over Spring REST Docs configuration than offered by
 include::{docs-java}/features/testing/springbootapplications/autoconfiguredspringrestdocs/withrestassured/MyRestDocsConfiguration.java[]
 ----
 
+If you want to make use of Spring REST Docs support for a parameterized output directory, you can create a `RestDocumentationFilter` bean.
+The auto-configuration calls `addFilter` with this filter, thereby causing each `RestAssured` call to automatically generate the default snippets.
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-docs/src/docs/asciidoc/features/testing.adoc:783]: This isn't true, unfortunately. There's no auto-configuration for REST Assured in Spring Boot at the moment. Can you please revert the REST Assured part of these changes?

File: spring-boot-project/spring-boot-docs/src/docs/asciidoc/features/testing.adoc [hunk 2]
```diff
@@ -751,6 +751,15 @@ If you require more control over Spring REST Docs configuration than offered by
 include::{docs-java}/features/testing/springbootapplications/autoconfiguredspringrestdocs/withwebtestclient/MyRestDocsConfiguration.java[]
 ----
 
+If you want to make use of Spring REST Docs support for a parameterized output directory, you can create a `WebTestClientBuilderCustomizer` bean.
+The auto-configuration calls `entityExchangeResultConsumer` with this consumer, thereby causing each `WebTestClient` call to automatically generate the default snippets.
```
Comment (wilkinsona) [inline spring-boot-project/spring-boot-docs/src/docs/asciidoc/features/testing.adoc:755]: This isn't quite right. The proposed code is correct so this description needs to be updated to match. The auto-configuration for `WebTestClient` calls the `WebTestClientBuilderCustomizer`. It's the customizer that then calls `entityExchangeResultConsumer` to set up the parameterized output.

**Review by wilkinsona — CHANGES_REQUESTED** [review]
Thanks very much for the PR, @brneto. I've left a couple of comments for your consideration when you have a moment.

File: spring-boot-project/spring-boot-docs/src/docs/asciidoc/features/testing.adoc (hunk 2 above)
Comment (brneto) [reply spring-boot-project/spring-boot-docs/src/docs/asciidoc/features/testing.adoc:755]: @wilkinsona what do you think of the following text:

> The auto-configuration calls this `WebTestClientBuilderCustomizer` in order to create the WebTestClient bean for the current context with the consumer provided in the parameter of the method `entityExchangeResultConsumer`.

File: spring-boot-project/spring-boot-docs/src/docs/asciidoc/features/testing.adoc (hunk 1 above)
Comment (brneto) [reply spring-boot-project/spring-boot-docs/src/docs/asciidoc/features/testing.adoc:783]: ok!

File: spring-boot-project/spring-boot-docs/src/docs/asciidoc/features/testing.adoc (hunk 2 above)
Comment (wilkinsona) [reply spring-boot-project/spring-boot-docs/src/docs/asciidoc/features/testing.adoc:755]: Looks good to me. Thanks.

**Comment by wilkinsona** [comment]
> It failed with message A build scan was not published as you have not authenticated with server 'ge.spring.io'., is there something more I have to do?

That's not a failure. It's just indicating that it couldn't publish a build scan to `ge.spring.io`. The build itself should be unaffected.

File: spring-boot-project/spring-boot-docs/src/docs/asciidoc/features/testing.adoc (hunk 2 above)
Comment (brneto) [reply spring-boot-project/spring-boot-docs/src/docs/asciidoc/features/testing.adoc]: commit pushed d3c6997698b5e7a0f04ced8b702a0a83d0a1af1b

**Comment by brneto** [comment]
I'm still intrigued why this doesn't work:
```java
@WebFluxTest
@AutoConfigureRestDocs
abstract class ContractTest {

    @TestConfiguration(proxyBeanMethods = false)
    static class RestDocsParameterizedOutput {

        @Bean
        WebTestClientBuilderCustomizer restDocsParameterizedOutput() {
            return builder -> builder.entityExchangeResultConsumer(document("{class-name}/{method-name}"));
        }
    }

    @Autowired
    private WebTestClient webTestClient;

    void webTestClientSetup() {
        RestAssuredWebTestClient.webTestClient(webTestClient);
    }
}
```

While this does:
```java
@WebFluxTest
@AutoConfigureRestDocs
abstract class ContractTest {

    @Autowired
    private ApplicationContext context;

    @Autowired
    private WebTestClientRestDocumentationConfigurer configurer;

    void webTestClientSetup() {
        RestAssuredWebTestClient.webTestClient(
                WebTestClient
                        .bindToApplicationContext(context)
                        .configureClient()
                        .filter(configurer)
                        .entityExchangeResultConsumer(document("{class-name}/{method-name}"))
                        .build());
    }
}
```

Any thought why this is happening?

Thanks!

**Comment by brneto** [comment]
@wilkinsona I've just figured out why this is happening! That's because the tests generated by the Spring Contract Test does not have the `@TestConfiguration` creating the `WebTestClientBuilderCustomizer` bean. Would you have any suggestion how I could overcome this? Thanks!

**Comment by brneto** [comment]
@wilkinsona It worked externalising the `@TestConfiguration` class and importing it.
e.g.
```java
@WebFluxTest
@AutoConfigureRestDocs
@Import(ContractTestConfiguration.class)
abstract class ContractTest {

    @Autowired
    private WebTestClient webTestClient;

    void webTestClientSetup() {
        RestAssuredWebTestClient.webTestClient(webTestClient);
    }
}
```

**Comment by brneto** [comment]
@wilkinsona is there something more I have to do or this is it?

**Comment by wilkinsona** [comment]
It's failing because the code in `MyWebTestClientBuilderCustomizerConfiguration` doesn't compile. It appears to be missing a static import for `WebTestClientRestDocumentation.document`. The package declaration is also incorrect. Don't worry about it, though. We can take care of it when we come to merge this.

**Comment by brneto** [comment]
Let me fix that. Give me a sec.

**Comment by brneto** [comment]
Just did 28eb7a011ae5c9587f9311b7edc6c774b49c9281. Sorry for that! :(

**Comment by brneto** [comment]
@wilkinsona it's still failing! Have I made any other mistake? :(

File: spring-boot-project/spring-boot-docs/src/main/java/org/springframework/boot/docs/features/testing/springbootapplications/autoconfiguredspringrestdocs/withwebtestclient/MyWebTestClientBuilderCustomizerConfiguration.java [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+import org.springframework.context.annotation.Bean;
+
+import static org.springframework.restdocs.webtestclient.WebTestClientRestDocumentation.document;
+
+@TestConfiguration(proxyBeanMethods = false)
+public class MyWebTestClientBuilderCustomizerConfiguration {
+
+	@Bean
+	public WebTestClientBuilderCustomizer restDocumentation() {
+		return builder -> builder.entityExchangeResultConsumer(document("{method-name}"));
```
Comment (scottfrederick) [inline spring-boot-project/spring-boot-docs/src/main/java/org/springframework/boot/docs/features/testing/springbootapplications/autoconfiguredspringrestdocs/withwebtestclient/MyWebTestClientBuilderCustomizerConfiguration.java:30]: Spring Boot has a checkstyle rule that requires parens around lambda args, so this should be `return (builder) ->`

**Comment by scottfrederick** [comment]
@brneto Just a small checkstyle violation. You can push another commit to fix that, or as Andy said we're happy to fix it when merging.

File: spring-boot-project/spring-boot-docs/src/main/java/org/springframework/boot/docs/features/testing/springbootapplications/autoconfiguredspringrestdocs/withwebtestclient/MyWebTestClientBuilderCustomizerConfiguration.java (hunk 3 above)
Comment (brneto) [reply spring-boot-project/spring-boot-docs/src/main/java/org/springframework/boot/docs/features/testing/springbootapplications/autoconfiguredspringrestdocs/withwebtestclient/MyWebTestClientBuilderCustomizerConfiguration.java:30]: I can do that. That's no problem! Give me a sec.

Comment (brneto) [reply spring-boot-project/spring-boot-docs/src/main/java/org/springframework/boot/docs/features/testing/springbootapplications/autoconfiguredspringrestdocs/withwebtestclient/MyWebTestClientBuilderCustomizerConfiguration.java]: Done: 7ad51d687b93221dc277f71ba449a9fa061bc72b

**Review by wilkinsona — APPROVED** [review]

**Comment by brneto** [comment]
@wilkinsona I'm migrating my project to Gradle and after that I started to receive this error message from my test:
```

    java.lang.NoClassDefFoundError: org/springframework/restdocs/ManualRestDocumentation
        at org.springframework.boot.test.autoconfigure.restdocs.RestDocumentationContextProviderRegistrar.registerBeanDefinitions(RestDocumentationContextProviderRegistrar.java:41)
        at org.springframework.context.annotation.ImportBeanDefinitionRegistrar.registerBeanDefinitions(ImportBeanDefinitionRegistrar.java:86)
        at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.lambda$loadBeanDefinitionsFromRegistrars$1(ConfigurationClassBeanDefinitionReader.java:396)
        at java.base/java.util.LinkedHashMap.forEach(LinkedHashMap.java:723)
        at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsFromRegistrars(ConfigurationClassBeanDefinitionReader.java:395)
        at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitionsForConfigurationClass(ConfigurationClassBeanDefinitionReader.java:157)
        at org.springframework.context.annotation.ConfigurationClassBeanDefinitionReader.loadBeanDefinitions(ConfigurationClassBeanDefinitionReader.java:129)
        at org.springframework.context.annotation.ConfigurationClassPostProcessor.processConfigBeanDefinitions(ConfigurationClassPostProcessor.java:343)
        at org.springframework.context.annotation.ConfigurationClassPostProcessor.postProcessBeanDefinitionRegistry(ConfigurationClassPostProcessor.java:247)
        at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanDefinitionRegistryPostProcessors(PostProcessorRegistrationDelegate.java:311)
        at org.springframework.context.support.PostProcessorRegistrationDelegate.invokeBeanFactoryPostProcessors(PostProcessorRegistrationDelegate.java:112)
        at org.springframework.context.support.AbstractApplicationContext.invokeBeanFactoryPostProcessors(AbstractApplicationContext.java:746)
        at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:564)
```
```
        Caused by:
        java.lang.ClassNotFoundException: org.springframework.restdocs.ManualRestDocumentation
            at java.base/jdk.internal.loader.BuiltinClassLoader.loadClass(BuiltinClassLoader.java:636)
            at java.base/jdk.internal.loader.ClassLoaders$AppClassLoader.loadClass(ClassLoaders.java:182)
            at java.base/java.lang.ClassLoader.loadClass(ClassLoader.java:519)
            ... 79 more
```

I haven't change anything in code but still I having this error with `qradle cleanTest test`. Any suggestion?

**Comment by wilkinsona** [comment]
@brneto A Spring Boot pull request isn’t a good place to ask for some support with Spring REST Docs. Please ask on Stack Overflow or [Gitter](https://gitter.im/spring-projects/spring-restdocs) instead.

**Comment by wilkinsona** [comment]
@brneto Thanks very much for making your first contribution to Spring Boot.

**Pull request closed without merging**